Feat: Adding some Item Components

Chaosgh opened the pull request:
Adds several ItemComponent:

CustomModelData
JukeboxPlayable
PersistentDataContainer
TooltipStyle

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Rich custom-model-data system with pluggable types (legacy int, boolean/float/string/color/byte/float/boolean arrays) and delegation.
  * New item components: model, tooltip style, jukebox playback, custom model data, and persistent-data-keyed values.
  * Expanded PDC support: byte/short/int/long/float/double, arrays (int/long/byte), boolean, and string.
* **Bug Fixes**
  * Version-aware item name matching for cross-version compatibility.
* **Other**
  * Amount matching added for item quantity checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Chaosgh committed:
added some item components

Chaosgh committed:
added some item components

Chaosgh committed:
fixed default mismatch

Chaosgh committed:
refactor ItemComponents Add InteractionContext for parity and added blank/empty checks.

Chaosgh committed:
Better handeling of invalide namespaces in ItemPDCComponent (for example white spaces or illegal chars)

Chaosgh committed:
Merge branch 'develop' into develop

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemCustomModelDataComponent.kt:
```diff
@@ -0,0 +1,47 @@
+class ItemCustomModelDataComponent(
+    @InnerMin(Min(0))
+    @Default("0")
+
```
Empty line should be removed.

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemCustomModelDataComponent.kt:
```diff
@@ -0,0 +1,47 @@
+            if (component.floats == expectedFloats) return@editMeta
+            component.floats = expectedFloats
+            meta.setCustomModelDataComponent(component)
+        }
```
Two things on this. 1. Check that this also compiles and runs on 1.21.3 as to not break version compatibility. If it doesn't work, then you need to use `serverVersion` (from packetevents see other uses) to check for compatibility 2. The input should be a list of floats instead of a singular int. Because we want to make this future compatible from the start. Otherwise if someone does need the floats, they are out of luck for a very long time. Make `customModelData: List<Var<Float>>`.

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemPDCComponent.kt:
```diff
@@ -0,0 +1,63 @@
+            actualValue == expected
+        }
+    }
+}
```
This either needs to support all non-recursive types, or be removed entirely.

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemTooltipStyleComponent.kt:
```diff
@@ -0,0 +1,39 @@
+        val expectedKey = key.get(player, interactionContext) ?: return false
+        val actualStyle = item.itemMeta?.tooltipStyle
+        val expectedNamespacedKey = NamespacedKey(expectedNamespace, expectedKey)
+        return actualStyle == expectedNamespacedKey
```
I don't think this works in pre‑1.21.5. If that is the case, we should add a check, and if the user is trying to do this in pre‑1.21.5, we should throw a clear error stating that it’s unsupported in their version.

gabber235 reviewed (changes requested):
We already discussed why some of these changes are important in vc. These are just as a reminder to what needed to be check/changed.

Chaosgh committed:
Merge branch 'develop' into develop

Chaosgh committed:
Small fix in `ItemNameComponent`: `effectiveName()` is only available since 1.21.4.

Chaosgh committed:
Merge remote-tracking branch 'origin/develop' into develop

Chaosgh commented:
> Is it possible to add the ItemModel component?

Shouldn't be too hard, i will need to look up if its in 1.21.3

Chaosgh committed:
Merge pull request #1 from gabber235/develop

r

Chaosgh committed:
- Data types were added to 'pdc' and 'CustomModelDataType'.

- Changed all namespace keys to a single string for user-friendliness.

- Added an item model component.

Chaosgh committed:
- Data types were added to 'pdc' and 'CustomModelDataType'.

- Changed all namespace keys to a single string for user-friendliness.

- Added an item model component.

Chaosgh committed:
small fixes to stay compatible with the typewriter panel

Chaosgh committed:
Merge pull request #2 from Chaosgh/components

Components

Chaosgh committed:
- different unsupported message to make it clearer
- now just uses pure shorts and longs

Chaosgh committed:
small cleanups

Chaosgh committed:
Merge branch 'develop' into develop

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemModelComponent.kt:
```diff
@@ -0,0 +1,47 @@
+import org.bukkit.entity.Player
+import org.bukkit.inventory.ItemStack
+
+@AlgebraicTypeInfo("item_model", Colors.BLUE, "fa6-solid:layer-group")
+class ItemModelComponent(
+    val modelKey: Var<String> = ConstVar("")
+) : ItemComponent {
+
+    override fun apply(player: Player?, interactionContext: InteractionContext?, item: ItemStack) {
+        item.editMeta { meta ->
+            val raw = modelKey.get(player, interactionContext)?.trim().orEmpty()
+            if (raw.isEmpty()) {
+                meta.itemModel = null
+                return@editMeta
+            }
+
+            val key = NamespacedKey.fromString(raw)
+            if (key == null) {
+                meta.itemModel = null
+                return@editMeta
+            }
+
+            meta.itemModel = key
+        }
+    }
+
+    override fun matches(player: Player?, interactionContext: InteractionContext?, item: ItemStack): Boolean {
+        val expectedRaw = modelKey.get(player, interactionContext)?.trim().orEmpty()
+        val actual = item.itemMeta?.itemModel
+
+        if (expectedRaw.isEmpty()) {
+            return actual == null
+        }
+
+        val expected = NamespacedKey.fromString(expectedRaw) ?: return actual == null
+        return actual == expected
+    }
+}
```
@Chaosgh I assume that code rabbit here is wrong and this works on 1.21.3 right? You checked this?

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemNameComponent.kt:
```diff
@@ -30,6 +32,12 @@ class ItemNameComponent(
+            item.effectiveName().plainText() == name.parsePlaceholders(player).stripped()
+        } else {
+            item.displayName().plainText() == name.parsePlaceholders(player).stripped()
+        }
```
The latest supported version is 1.21.3, so I think this is either the wrong version, or just not necessary.

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemPDCComponent.kt:
```diff
@@ -0,0 +1,38 @@
+
+@AlgebraicTypeInfo("persistent_data_container", Colors.PURPLE, "fa6-solid:database")
+class ItemPDCComponent(
+
```
Small thing, but please remove this empty line

gabber235 commented on engine/engine-paper/src/main/kotlin/com/typewritermc/engine/paper/utils/item/components/ItemTooltipStyleComponent.kt:
```diff
@@ -0,0 +1,50 @@
+
+@AlgebraicTypeInfo("tooltip_style", Colors.CYAN, "fa6-solid:palette")
+class ItemTooltipStyleComponent(
+
```
Small thing, but please remove this empty line

gabber235 reviewed (changes requested):
Looks pretty good already, some small changes still need to be made.

Chaosgh committed:
small cleanups like gabber told me

Chaosgh committed:
small cleanups like gabber told me

Chaosgh committed:
Merge branch 'develop' into develop

Chaosgh committed:
Merge remote-tracking branch 'origin/develop' into develop

Chaosgh commented:
Sorry btw that i was so fast in the call, there was a private emergency i had

Chaosgh committed:
Merge branch 'develop' into develop

Chaosgh committed:
Merge branch 'develop' into develop
